`refrax` fails on real projects

when I run `refrax refactor . --ai=deepseek`, the refactoring proceeds successfully, however at the end I get the following error:

```
5:19PM INF critic: critic server stopped successfully
panic: failed to close resource: failed to stop fixer server: context deadline exceeded
        panic: failed to close resource: failed to stop facilitator server: context deadline exceeded

goroutine 1 [running]:
github.com/cqfn/refrax/internal/client.closeResource({0x10620be0?, 0xc000020e60?})
        /Users/lombrozo/Workspace/OpenSource/refrax/internal/client/refrax_client.go:189 +0x79
panic({0x1058b540?, 0xc000032bd0?})
        /usr/local/Cellar/go/1.24.4/libexec/src/runtime/panic.go:792 +0x132
github.com/cqfn/refrax/internal/client.closeResource({0x10620c00?, 0xc000116300?})
        /Users/lombrozo/Workspace/OpenSource/refrax/internal/client/refrax_client.go:189 +0x79
github.com/cqfn/refrax/internal/client.(*RefraxClient).Refactor(0xc0005adb10, {0x10620b60, 0xc0000327d0})
        /Users/lombrozo/Workspace/OpenSource/refrax/internal/client/refrax_client.go:148 +0x130a
github.com/cqfn/refrax/internal/client.Refactor(0xc00017e1b0)
        /Users/lombrozo/Workspace/OpenSource/refrax/internal/client/refrax_client.go:41 +0x153
github.com/cqfn/refrax/cmd.newRefactorCmd.func1(0xc000186200?, {0xc00018c3c0?, 0x4?, 0x1017e607?})
        /Users/lombrozo/Workspace/OpenSource/refrax/cmd/refactor.go:25 +0x93
github.com/spf13/cobra.(*Command).execute(0xc00018a308, {0xc00018c360, 0x3, 0x3})
        /Users/lombrozo/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1015 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0xc00018a008)
        /Users/lombrozo/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x46f
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/lombrozo/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071
github.com/cqfn/refrax/cmd.Execute()
        /Users/lombrozo/Workspace/OpenSource/refrax/cmd/root.go:13 +0x30
main.main()
        /Users/lombrozo/Workspace/OpenSource/refrax/main.go:12 +0x13
```

Most probably it's some issue with incorrectly closed resources.  We need to investigate the root cause of the issue.

Original problem is connected with timeouts:

```
panic: failed to close resource: failed to stop fixer server: context deadline exceeded
        panic: failed to close resource: failed to stop facilitator server: context deadline exceeded
```

--- a/internal/critic/server.go
+++ b/internal/critic/server.go
@@ -1,6 +1,7 @@
 package critic
 
 import (
+	"context"
 	"encoding/base64"
 	"fmt"
 	"net/http"
@@ -79,7 +80,16 @@ func (c *Critic) Handler(handler protocol.Handler) {
 	c.server.Handler(handler)
 }
 
-func (c *Critic) think(m *protocol.Message) (*protocol.Message, error) {
+func (c *Critic) think(ctx context.Context, m *protocol.Message) (*protocol.Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, fmt.Errorf("context canceled: %w", ctx.Err())
+	default:
+		return c.thinkLong(m)
+	}
+}
+
+func (c *Critic) thinkLong(m *protocol.Message) (*protocol.Message, error) {
 	c.log.Debug("received message: #%s", m.MessageID)
 	var java string
 	var task string

--- a/internal/critic/server_test.go
+++ b/internal/critic/server_test.go
@@ -1,6 +1,7 @@
 package critic
 
 import (
+	"context"
 	"errors"
 	"testing"
 
@@ -113,7 +114,7 @@ func TestCriticThink_ReturnsMessage(t *testing.T) {
 		MessageID("msg-123").
 		Build()
 
-	response, err := critic.think(msg)
+	response, err := critic.think(context.Background(), msg)
 
 	require.NoError(t, err)
 	assert.Equal(t, msg.MessageID, response.MessageID)

--- a/internal/facilitator/server.go
+++ b/internal/facilitator/server.go
@@ -3,6 +3,7 @@
 package facilitator
 
 import (
+	"context"
 	"encoding/base64"
 	"fmt"
 	"net/http"
@@ -65,7 +66,16 @@ func (f *Facilitator) Handler(handler protocol.Handler) {
 	f.server.Handler(handler)
 }
 
-func (f *Facilitator) think(m *protocol.Message) (*protocol.Message, error) {
+func (f *Facilitator) think(ctx context.Context, m *protocol.Message) (*protocol.Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, fmt.Errorf("context canceled: %w", ctx.Err())
+	default:
+		return f.thinkLong(m)
+	}
+}
+
+func (f *Facilitator) thinkLong(m *protocol.Message) (*protocol.Message, error) {
 	f.log.Debug("received message: #%s", m.MessageID)
 	var file *protocol.FilePart
 	var task string
@@ -92,7 +102,6 @@ func (f *Facilitator) think(m *protocol.Message) (*protocol.Message, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to ask critic: %w", err)
 	}
-
 	criticMessage := criticResp.Result.(protocol.Message)
 	var suggestions []string
 	for _, part := range criticMessage.Parts {

--- a/internal/fixer/server.go
+++ b/internal/fixer/server.go
@@ -2,6 +2,7 @@
 package fixer
 
 import (
+	"context"
 	"encoding/base64"
 	"fmt"
 	"net/http"
@@ -46,33 +47,59 @@ func NewFixer(ai brain.Brain, port int) *Fixer {
 }
 
 // Start begins the Fixer server and signals readiness through the provided channel.
-func (c *Fixer) Start(ready chan<- struct{}) error {
-	c.log.Info("starting fixer server on port %d...", c.port)
+func (f *Fixer) Start(ready chan<- struct{}) error {
+	f.log.Info("starting fixer server on port %d...", f.port)
 	var err error
-	if err = c.server.Start(ready); err != nil && err != http.ErrServerClosed {
+	if err = f.server.Start(ready); err != nil && err != http.ErrServerClosed {
 		return fmt.Errorf("failed to start fixer server: %w", err)
 	}
 	return err
 }
 
 // Close gracefully stops the Fixer server.
-func (c *Fixer) Close() error {
-	c.log.Info("stopping fixer server...")
-	if err := c.server.Close(); err != nil {
+func (f *Fixer) Close() error {
+	f.log.Info("stopping fixer server...")
+	if err := f.server.Close(); err != nil {
 		return fmt.Errorf("failed to stop fixer server: %w", err)
 	}
-	c.log.Info("fixer server stopped successfully")
+	f.log.Info("fixer server stopped successfully")
 	return nil
 }
 
 // Handler sets the handler function for processing requests on the Fixer server.
-func (c *Fixer) Handler(hander protocol.Handler) {
-	c.server.Handler(hander)
+func (f *Fixer) Handler(hander protocol.Handler) {
+	f.server.Handler(hander)
 }
 
-func (c *Fixer) think(m *protocol.Message) (*protocol.Message, error) {
-	c.log.Info("received message: #%s", m.MessageID)
-	c.log.Info("trying to fix Java code...")
+func (f *Fixer) think(ctx context.Context, m *protocol.Message) (*protocol.Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, fmt.Errorf("context canceled: %w", ctx.Err())
+	case res := <-f.thinkChan(m):
+		return res.msg, res.err
+	}
+}
+
+type thought struct {
+	msg *protocol.Message
+	err error
+}
+
+func (f *Fixer) thinkChan(m *protocol.Message) <-chan thought {
+	res := make(chan thought, 1)
+	go func() {
+		msg, err := f.thinkLong(m)
+		res <- thought{
+			msg, err,
+		}
+		close(res)
+	}()
+	return res
+}
+
+func (f *Fixer) thinkLong(m *protocol.Message) (*protocol.Message, error) {
+	f.log.Info("received message: #%s", m.MessageID)
+	f.log.Info("trying to fix Java code...")
 	var code string
 	var suggestions []string
 	var class string
@@ -97,14 +124,14 @@ func (c *Fixer) think(m *protocol.Message) (*protocol.Message, error) {
 	}
 	all := strings.Join(suggestions, "\n")
 	question := fmt.Sprintf(prompt, class, code, all)
-	c.log.Info("asking AI to fix java code...")
-	c.log.Debug("asking AI: %s", question)
-	answer, err := c.brain.Ask(question)
+	f.log.Info("asking AI to fix java code...")
+	f.log.Debug("asking AI: %s", question)
+	answer, err := f.brain.Ask(question)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get answer from AI: %w", err)
 	}
-	c.log.Debug("received answer from AI: %s", answer)
-	c.log.Info("AI provided a fix for the Java code, sending response back...")
+	f.log.Debug("received answer from AI: %s", answer)
+	f.log.Info("AI provided a fix for the Java code, sending response back...")
 	message := protocol.NewMessageBuilder().
 		MessageID(m.MessageID).
 		Part(protocol.NewFileBytes([]byte(clean(answer)))).

--- a/internal/protocol/custom_server.go
+++ b/internal/protocol/custom_server.go
@@ -24,7 +24,7 @@ type (
 	// Handler handles all incoming requests on JSONRPC level
 	Handler func(next Handler, r *JSONRPCRequest) (*JSONRPCResponse, error)
 	// MsgHandler handles messages received from the A2A server on Message level
-	MsgHandler func(message *Message) (*Message, error)
+	MsgHandler func(ctx context.Context, message *Message) (*Message, error)
 )
 
 // NewCustomServer creates a new instance of a custom server that handles A2A requests
@@ -71,12 +71,14 @@ func (serv *customServer) Start(ready chan<- struct{}) error {
 	return err
 }
 
-// Close stops the custom server gracefully, allowing for a timeout.
+// Close stops the custom server gracefully, allowing for a timeout.ffile.go
+// @todo #75:60min Shutdown should be used instead of Close to allow for graceful shutdown.
+// This will allow the server to finish processing ongoing requests before stopping.
+// Cruntly, Close is used because of:
+// https://stackoverflow.com/questions/79714562/how-to-cancel-long-running-tasks-when-a-go-http-server-is-shut-down
 func (serv *customServer) Close() error {
 	log.Debug("stopping custom a2a server on port %d...", serv.port)
-	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
-	defer cancel()
-	return serv.server.Shutdown(ctx)
+	return serv.server.Close()
 }
 
 func (serv *customServer) handleAgentCard(w http.ResponseWriter, r *http.Request) {
@@ -113,9 +115,9 @@ func (serv *customServer) handleJSONRPC(w http.ResponseWriter, r *http.Request) 
 	var err error
 	if serv.handler != nil {
 		start := serv.handler
-		resp, err = start(basic(serv.msgHandler), &req)
+		resp, err = start(basic(r.Context(), serv.msgHandler), &req)
 	} else {
-		start := basic(serv.msgHandler)
+		start := basic(r.Context(), serv.msgHandler)
 		resp, err = start(nil, &req)
 	}
 	if err != nil {
@@ -125,7 +127,7 @@ func (serv *customServer) handleJSONRPC(w http.ResponseWriter, r *http.Request) 
 	return send(w, resp)
 }
 
-func basic(mh MsgHandler) Handler {
+func basic(ctx context.Context, mh MsgHandler) Handler {
 	return func(_ Handler, r *JSONRPCRequest) (*JSONRPCResponse, error) {
 		id := str(r.ID)
 		switch r.Method {
@@ -144,7 +146,7 @@ func basic(mh MsgHandler) Handler {
 			}
 			log.Debug("handling JSON-RPC request: %s, params: %v", r.Method, params)
 			msg := params.Message
-			msg, err = mh(msg)
+			msg, err = mh(ctx, msg)
 			if err != nil {
 				resp := failure(id, ErrCodeInternalError, fmt.Sprintf("failed to handle message send: %v", err))
 				return &resp, nil
@@ -205,7 +207,7 @@ func send(w http.ResponseWriter, r *JSONRPCResponse) error {
 	return json.NewEncoder(w).Encode(r)
 }
 
-func record(message *Message) (*Message, error) {
+func record(_ context.Context, message *Message) (*Message, error) {
 	log.Debug("server received the following message: %s", message.MessageID)
 	return message, nil
 }

--- a/internal/protocol/custom_server_test.go
+++ b/internal/protocol/custom_server_test.go
@@ -2,6 +2,7 @@ package protocol
 
 import (
 	"bytes"
+	"context"
 	"encoding/json"
 	"fmt"
 	"net/http"
@@ -88,7 +89,7 @@ func testServer(t *testing.T) (server Server, port int, ready chan struct{}) {
 	return server, port, ready
 }
 
-func joke(msg *Message) (*Message, error) {
+func joke(_ context.Context, msg *Message) (*Message, error) {
 	log.Debug("Received message: %s", msg.MessageID)
 	if len(msg.Parts) == 0 || msg.Parts[0].PartKind() != PartKindText || msg.Parts[0].(*TextPart).Text != "tell me a joke" {
 		return nil, fmt.Errorf("unexpected message content, we expected 'tell me a joke', got: '%v'", msg.Parts[0].(*TextPart).Text)

